### PR TITLE
Add the req_id attribute on the decision logs

### PR DESCRIPTION
**Pull request by humbertoc-silva**
Today it is not possible to correlate the decision log with other types of logs (server, print, etc.) when the server log level is >= INFO. The log correlation could be helpful in troubleshooting.

A solution is to add a common attribute in all logs to make the log correlation possible, so adding the req_id attribute on decision logs, when server log level is >= INFO, will make it possible.

Fixes: #5006

Signed-off-by: Humberto Corrêa da Silva <humbertoc_silva@hotmail.com>

### CODE REVIEW

File: logging/logging_test.go [hunk 1]
```diff
@@ -82,3 +82,47 @@ func TestWithFieldsMerges(t *testing.T) {
...
+		ReqID:      1,
+		ReqMethod:  "GET",
+		ReqPath:    "/test",
+	}.Fields()
+
+	var fieldvalue interface{}
+	var ok bool
+
+	if fieldvalue, ok = fields["client_addr"]; !ok {
+		t.Fatal("Fields did not contain the client_addr field")
```
Comment (anderseknert) [inline logging/logging_test.go:98]: I think we can use Error rather than Fatal for these assertions, no?

**Review by anderseknert — COMMENTED** [review]
Code looks good to me, but I'll defer to @srenatus for the architectural review. One thing I don't understand though — where/when is the actual request ID created? I only see it being referenced. Has that been done elsewhere? 😅

**Comment by anderseknert** [comment]
Oh, I just realized — there is no documentation for this feature included in this PR. Since this is something new that will show up in logs, and decision logs, we'll need to document what the purpose of the field is.

**Comment by humbertoc-silva** [comment]
@anderseknert I will add the documentation in the next commit and change `Fatal` to `Error` too.

The request ID is generated here: https://github.com/humbertoc-silva/opa/blob/decision-log-req-id/runtime/logging.go#L58

**Comment by anderseknert** [comment]
I see! Using an integer and auto-increment seems like it would risk making this less useful, or no? If you have 100 OPAs in a cluster, they are all going to have a "request_id": 0, 1, 2, 3, 4, 5, 6 ... etc, in their logs, and the counter is of course reset whenever an instance is restarted, so after a while you are going to see _thousands_ of "request_id": 0 in your log aggregator... and if the purpose of this was to help correlate a request with other identifiers, would this help much with that? I know that wasn't part of your PR, just asking so it doesn't turn out to be unusable for what you intended it for 😅

**Comment by humbertoc-silva** [comment]
Maybe not. Using de example of 100 OPA instances:

Instance 1 - Instance ID: 123
req_id: 1 (request, response, print, decision_log)
req_id: 2 (request, response, print, decision_log)
and so on...

Instance 2 - Instance ID: 456
req_id: 1 (request, response, print, decision_log)
req_id: 2 (request, response, print, decision_log)
and so on...

Today we can use some information related to the instance, the ID for example, to aggregate the logs by instance. If an instance reset I will consider the new instance ID and the the request ID counting start again, as expected.

To avoid this type of problem, the OPA request ID could be an unique value (GUID), so we wouldn't depend on extra values to aggregate the information. This makes sense for you?

I do not get this part:

> so after a while you are going to see thousands of "request_id": 0 in your log aggregator...

When the request ID would be equals 0.

**Comment by srenatus** [comment]
Sorry I haven't gotten to this yet. I'll make sure to review this a bit more this week.

**Comment by humbertoc-silva** [comment]
I was thinking that would be interesting if we had a new field on logs, e.g `correlation_id`, dedicated for this purpose. The value could be a unique value (GUID), by default OPA could generate a value but this could be configurable too.

Example:

- Default: OPA will generate a unique value on each request.
- Via configuration: OPA will read a specific header (e.g X-Correlation-ID, X-Request-ID, etc.) and use it on all logs so that we could correlate external logs too.

Could this be a future improvement for this PR?

**Comment by srenatus** [comment]
> I was thinking that would be interesting if we had a new field on logs, e.g `correlation_id`, dedicated for this purpose or continue to use `req_id`. The value could be a unique value (GUID), by default OPA could generate a value but this could be configurable too.
> 
> Example:
> 
> * Default: OPA will generate a unique value on each request.
> * Via configuration: OPA will read a specific header (e.g X-Correlation-ID, X-Request-ID, etc.) and use it on all logs so that we could correlate external logs too.
> 
> Could this be a future improvement for this PR?

There have been related issues, I think (https://github.com/open-policy-agent/opa/issues/5008); I think what you propose there does make sense. I'm not certain about the potential overlap though -- there's a lot going on on OpenTelemetry, too, with span and trace id #5230... Let's get that discussion going in one of those issues, or a new one.

File: docs/content/management-decision-logs.md [hunk 2]
```diff
@@ -76,6 +76,7 @@ Decision log updates contain the following fields:
 | `[_].erased` | `array[string]` | Set of JSON Pointers specifying fields in the event that were erased. |
 | `[_].masked` | `array[string]` | Set of JSON Pointers specifying fields in the event that were masked. |
 | `[_].nd_builtin_cache` | `object` | Key-value pairs of non-deterministic builtin names, paired with objects specifying the input/output mappings for each unique invocation of that builtin during policy evaluation. Intended for use in debugging and decision replay. Receivers will need to decode the JSON using Rego's JSON decoders. |
+| `[_].req_id` | `number` | Unique request identifier for the request that started the policy query. The attribute value is the same as the value present in others logs (request, response, and print) and could be used to correlate them all. This attribute will be included just when OPA runtime is initialized in server mode and the log level is equal or greater than info. |
```
Comment (srenatus) [inline docs/content/management-decision-logs.md:79]: [nit] "Unique" -- We should mention that it's incrementing, and unique only to the OPA instance. Not globally unique or some such thing.

File: logging/logging.go [hunk 3]
```diff
@@ -185,3 +185,27 @@ func (l *NoOpLogger) SetLevel(level Level) {
 func (l *NoOpLogger) GetLevel() Level {
 	return l.level
 }
+
+type requestContextKey string
+
+// ReqCtxKey is the key used to access the request context.
+const ReqCtxKey = requestContextKey("request-context-key")
```
Comment (srenatus) [inline logging/logging.go:192]: Exporting this is defeating the purpose here: It's done via non-exported type to ensure that we control how it's set and read.

Alternatively, we could keep `reqCtxKey` unexported and expose a method like
```go
func WithRequestID(ctx context.Context, reqID string) context.Context
```
that injects the request ID accordingly, and
```go
func RequestIDFromContext(context.Context) string
```
or something like that for retrieval.

**Review by srenatus — COMMENTED** [review]
Sorry it took me so long to review this -- please have a look at the inline comment 👇

**Comment by humbertoc-silva** [comment]
> > I was thinking that would be interesting if we had a new field on logs, e.g `correlation_id`, dedicated for this purpose or continue to use `req_id`. The value could be a unique value (GUID), by default OPA could generate a value but this could be configurable too.
> > Example:
> > 
> > * Default: OPA will generate a unique value on each request.
> > * Via configuration: OPA will read a specific header (e.g X-Correlation-ID, X-Request-ID, etc.) and use it on all logs so that we could correlate external logs too.
> > 
> > Could this be a future improvement for this PR?
> 
> There have been related issues, I think (#5008); I think what you propose there does make sense. I'm not certain about the potential overlap though -- there's a lot going on on OpenTelemetry, too, with span and trace id #5230... Let's get that discussion going in one of those issues, or a new one.

Ok, @srenatus
I need to start some tests with OPA and OpenTelemetry to see how things work to contribute to discussions.

File: docs/content/management-decision-logs.md (hunk 2 above)
Comment (humbertoc-silva) [reply docs/content/management-decision-logs.md]: Done.

File: logging/logging.go (hunk 3 above)
Comment (humbertoc-silva) [reply logging/logging.go]: Done.

**Review by srenatus — APPROVED** [review]
Looks good to me. WDYT @anderseknert?

**Review by anderseknert — APPROVED** [review]
LGTM 👍

**Pull request merged**